### PR TITLE
Allow pasting shortcodes in blocks with content.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
Until now when pasting a shortcode in existing blocks if the text we paste is plain text the inline pasting mechanism takes place and the shortcode is not parsed and the text is pasted as normal plain text. If the text contains markup the shortcode is parsed.
So if we paste [gallery ids="ID"] in a paragraph with content the text is appended to the paragraph (If copying from here span's are added so this does not happen).
If we paste `<div>[gallery ids="ID"]</div>` in a paragraph with content the shortcode is parsed and an equivalent block is added.
This change makes sure inline pasting mechanism does not take place if pasting shortcodes.

## How Has This Been Tested?
Add some content to a paragraph or list or heading paste a shortcode verify it is correctly parsed and a block equivalent to the shortcode is added.
Try paste other content besides shortcodes and verify things work as before.

## Screenshots (jpeg or gifs if applicable):

Before: 
dec-08-2017 21-45-16

After:
dec-08-2017 21-43-45

### CODE REVIEW

File: blocks/api/raw-handling/index.js [hunk 1]
```diff
@@ -57,8 +57,14 @@ export default function rawHandler( { HTML, plainText = '', mode = 'AUTO' } ) {
 		HTML = converter.makeHtml( plainText );
 	}
 
+	const shortcodesConverted = shortcodeConverter( HTML );
```
Comment (mcsf) [inline blocks/api/raw-handling/index.js]: The morphology (esp. the ending in an adjective) irks me a bit. Perhaps renaming and adding a comment:

```js
// An array of HTML strings and block objects. The blocks replace matched shortcodes.
const pieces = …;
```

An alternative could be the quite awkward `htmlAndBlocks`. :)

File: blocks/api/raw-handling/index.js [hunk 2]
```diff
@@ -57,8 +57,14 @@ export default function rawHandler( { HTML, plainText = '', mode = 'AUTO' } ) {
...
+
+	//if the shortcodeConverter returns just one element it has no shortcodes
+	//even if we paste only one shortcode it returns more than one element because empty strings are added
```
Comment (mcsf) [inline blocks/api/raw-handling/index.js]: Comment format should be:

```
// If the …
// even if
```

i.e., capitalized sentences and a leading space for each line. Bonus if lines are limited in length (my preference is 78/80 chars, but commonly we see 100 as a hard limit).

Comment (mcsf) [reply blocks/api/raw-handling/index.js]: As for the comment itself, I'd make it terser, e.g. _shortcodeConverter will always return more than one element if shortcodes are matched, as empty HTML strings are included_.

File: blocks/api/raw-handling/index.js [hunk 3]
```diff
@@ -57,8 +57,14 @@ export default function rawHandler( { HTML, plainText = '', mode = 'AUTO' } ) {
...
+	const hasShortCodes = shortcodesConverted.length > 1;
```
Comment (mcsf) [inline blocks/api/raw-handling/index.js]: Spelling: `hasShortcodes`

File: blocks/api/raw-handling/index.js [hunk 4]
```diff
@@ -57,8 +57,14 @@ export default function rawHandler( { HTML, plainText = '', mode = 'AUTO' } ) {
...
+
 	// Return filtered HTML if it's inline paste or all content is inline.
-	if ( mode === 'INLINE' || ( mode === 'AUTO' && isInlineContent( HTML ) ) ) {
+	if ( ! hasShortCodes && ( mode === 'INLINE' || ( mode === 'AUTO' && isInlineContent( HTML ) ) ) ) {
```
Comment (mcsf) [inline blocks/api/raw-handling/index.js]: Perhaps refactoring like so:

```js
const isInline = mode === 'INLINE' || ( mode === 'AUTO' && isInlineContent( HTML ) );
const hasShortcodes = …;
if ( ! hasShortcodes && isInline ) {
```

**Review by mcsf — CHANGES_REQUESTED** [review]
I recommend adding tests for this, as there's a lot going on during pasting in general. :)

**Comment by mcsf** [comment]
Referencing related issue #3806.

We can probably go ahead with this fix, but the logic in `raw-handler` is getting less obvious over time. Something to think about.

**Comment by jorgefilipecosta** [comment]
Hi @mcsf, thank you for your review I tried to address the points you raised. Regarding the tests, it looks like we don't have yet tests for the shortcode parsing mechanism. That mechanism uses window.wp.shortcode, which is not available during test execution. I'm not sure what is the best way to add test cases to it probably we will need to use some mocks.

**Comment by mcsf** [comment]
Yeah, I think it’s fine to mock it specifically for these tests.

**Review by mcsf — APPROVED** [review]

**Comment by mcsf** [comment]
I won't block this PR because of the tests, but we should get them in.

**Pull request merged**